File: app/src/test/java/com/matchthree/data/HighScoreStoreTest.kt
package com.matchthree.data

import androidx.datastore.preferences.core.PreferenceDataStoreFactory
import com.matchthree.ui.GameMode
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.cancel
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.runBlocking
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test
import java.io.File

/** JVM tests for the M5 high-score persistence logic. */
class HighScoreStoreTest {

    private fun tempStore(): Pair<HighScoreStore, CoroutineScope> {
        val file = File.createTempFile("high-scores-test", ".preferences_pb")
        file.deleteOnExit()
        val scope = CoroutineScope(Dispatchers.IO + Job())
        val store = HighScoreStore(
            PreferenceDataStoreFactory.create(scope = scope) { file },
        )
        return store to scope
    }

    @Test
    fun saveIfBeatsPersistsNewHighScore() = runBlocking {
        val (store, scope) = tempStore()
        try {
            assertTrue(store.saveIfBeats(GameMode.CLASSIC, 500))
            assertEquals(500, store.scores.first().classic)
        } finally {
            scope.cancel()
        }
    }

    @Test
    fun saveIfBeatsKeepsHighestScoreOnly() = runBlocking {
        val (store, scope) = tempStore()
        try {
            store.saveIfBeats(GameMode.CLASSIC, 500)
            assertFalse(store.saveIfBeats(GameMode.CLASSIC, 300))
            assertEquals(500, store.scores.first().classic)
            assertTrue(store.saveIfBeats(GameMode.CLASSIC, 900))
            assertEquals(900, store.scores.first().classic)
        } finally {
            scope.cancel()
        }
    }

    @Test
    fun classicAndZenScoresAreIndependent() = runBlocking {
        val (store, scope) = tempStore()
        try {
            store.saveIfBeats(GameMode.CLASSIC, 500)
            store.saveIfBeats(GameMode.ZEN, 900)
            val scores = store.scores.first()
            assertEquals(500, scores.classic)
            assertEquals(900, scores.zen)
        } finally {
            scope.cancel()
        }
    }

    @Test
    fun nonPositiveScoresAreNeverSaved() = runBlocking {
        val (store, scope) = tempStore()
        try {
            assertFalse(store.saveIfBeats(GameMode.ZEN, 0))
            assertFalse(store.saveIfBeats(GameMode.ZEN, -10))
            assertEquals(0, store.scores.first().zen)
        } finally {
            scope.cancel()
        }
    }

    @Test
    fun forModeMapsToTheRightPerModeField() {
        val scores = HighScores(classic = 100, zen = 200)
        assertEquals(100, scores.forMode(GameMode.CLASSIC))
        assertEquals(200, scores.forMode(GameMode.ZEN))
    }
}

File: Tests/HighScoreStoreTest.cs
using MatchThree.Engine.Data;
using NUnit.Framework;

namespace MatchThree.Engine.Tests;

/// <summary>JVM tests for the M5 high-score persistence logic (mirrors app/src/test HighScoreStoreTest).</summary>
public class HighScoreStoreTest
{
    private string _tempDir = null!;
    private HighScoreStore _store = null!;

    [SetUp]
    public void SetUp()
    {
        _tempDir = Path.Combine(Path.GetTempPath(), $"matchthree-tests-{Guid.NewGuid():N}");
        _store = new HighScoreStore(_tempDir);
    }

    [TearDown]
    public void TearDown()
    {
        if (Directory.Exists(_tempDir)) Directory.Delete(_tempDir, recursive: true);
    }

    [Test]
    public void SaveIfBeatsPersistsNewHighScore()
    {
        Assert.That(_store.SaveIfBeats(GameMode.Classic, 500), Is.True);
        Assert.That(_store.Load().Classic, Is.EqualTo(500));
    }

    [Test]
    public void SaveIfBeatsKeepsHighestScoreOnly()
    {
        _store.SaveIfBeats(GameMode.Classic, 500);
        Assert.That(_store.SaveIfBeats(GameMode.Classic, 300), Is.False);
        Assert.That(_store.Load().Classic, Is.EqualTo(500));
        Assert.That(_store.SaveIfBeats(GameMode.Classic, 900), Is.True);
        Assert.That(_store.Load().Classic, Is.EqualTo(900));
    }

    [Test]
    public void ClassicAndZenScoresAreIndependent()
    {
        _store.SaveIfBeats(GameMode.Classic, 500);
        _store.SaveIfBeats(GameMode.Zen, 900);
        var scores = _store.Load();
        Assert.That(scores.Classic, Is.EqualTo(500));
        Assert.That(scores.Zen, Is.EqualTo(900));
    }

    [Test]
    public void NonPositiveScoresAreNeverSaved()
    {
        Assert.That(_store.SaveIfBeats(GameMode.Zen, 0), Is.False);
        Assert.That(_store.SaveIfBeats(GameMode.Zen, -10), Is.False);
        Assert.That(_store.Load().Zen, Is.EqualTo(0));
    }

    [Test]
    public void CorruptFileReadsAsZeros()
    {
        Directory.CreateDirectory(_tempDir);
        File.WriteAllText(Path.Combine(_tempDir, "highscores.json"), "not json {{{");
        Assert.That(_store.Load(), Is.EqualTo(new HighScores()));
    }

    [Test]
    public void ForModeMapsToTheRightPerModeField()
    {
        var scores = new HighScores(Classic: 100, Zen: 200);
        Assert.That(scores.ForMode(GameMode.Classic), Is.EqualTo(100));
        Assert.That(scores.ForMode(GameMode.Zen), Is.EqualTo(200));
    }
}